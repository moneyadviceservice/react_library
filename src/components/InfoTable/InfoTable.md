### The InfoTable component enables you to display information in a custom table format.
You can either use the `textContent` prop to pass text to be displayed inside the table or for more complex scenarios you can just pass the contents as children. Table colors are also customizable.

```jsx
import React, { useState } from 'react';
import { InfoTable, Row } from '@moneypensionservice/directories';

const lorem = 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Etiam semper diam at erat pulvinar, at pulvinar felis blandit. Vestibulum volutpat tellus diam, consequat gravida libero rhoncus ut.';

<Row justify="space-around" direction={{xs: 'column', md: 'row'}}>
  <InfoTable 
    margin={{bottom: '15px'}}
    sizes={{xs: 12, md: 4}}
    title="Default InfoTable Component" 
    textContent={lorem} />
  <InfoTable 
    sizes={{xs: 12, md: 4}}
    title="I'm using custom colors!" 
    textContent={lorem}
    tableColor="aqua"
    titleColor="black" />
</Row>
```
