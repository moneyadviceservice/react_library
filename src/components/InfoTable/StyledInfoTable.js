import styled, { css } from 'styled-components'
import { Col } from '../Grid'
import { Paragraph } from '../Paragraph'

const TableContainer = styled(Col)`
  padding: 0;
  ${({ tableColor, theme }) =>
    css`
      border: 2px solid ${tableColor || theme.colors.infotable.default};
    `}
`

const TableIcon = styled.span`
  margin-right: 10px;
  & > * {
    height: 30px;

    ${({ titleColor }) =>
      css`
        stroke: ${titleColor};
      `}
  }
`

const TableHead = styled(Col)`
  font-size: 1.375rem;
  font-weight: 500;
  line-height: 1.4375rem;
  padding: 0.75rem;

  ${({ titleColor }) =>
    css`
      color: ${titleColor};
    `}
  ${({ tableColor, theme }) =>
    css`
      background-color: ${tableColor || theme.colors.infotable.default};
    `}
`

const TableContent = styled(Col)`
  font-size: 1rem;
  line-height: 1.4375rem;
  color: ${({ theme }) => theme.colors.infotable.content};
  ${({ padding }) =>
    !padding &&
    css`
      padding: 1.5rem;
    `}
`

const TableText = styled(Paragraph)``

export { TableContainer, TableIcon, TableHead, TableContent, TableText }
