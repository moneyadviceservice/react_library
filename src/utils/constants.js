export const breakpoints = {
  xs: 1,
  sm: 48, // 768px
  md: 64, // 1024px
  lg: 90, // 1440px
  xl: 120, // 1920px
}

export const dimensions = ['xs', 'sm', 'md', 'lg', 'xl']

export const mediaQuery = 'only screen'

export const ALIGN_SELF_MAP = {
  center: 'center',
  end: 'flex-end',
  start: 'flex-start',
  stretch: 'stretch',
}

export const SPACING_SIZES = [
  'xxsmall',
  'xsmall',
  'small',
  'medium',
  'large',
  'xlarge',
]

