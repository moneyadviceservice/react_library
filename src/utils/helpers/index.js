export * from './background'
export * from './responsive'
export * from './flexbox'
export * from './styles'
export * from './mediaQueries'
export * from './sizing'
export * from './refs'
