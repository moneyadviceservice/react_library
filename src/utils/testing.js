import { render, act } from '@testing-library/react'
import GlobalTheme from '../components/ThemeProvider'

const customRender = (ui, options) =>
  render(ui, { wrapper: GlobalTheme, ...options })

const waitRender = async (ui, options) => {
  const { container, ...rest } = customRender(ui, options)

  await act(async () => {
    await Promise.resolve(container)
  })

  return { container, ...rest }
}

export * from '@testing-library/react'

export { customRender as render, waitRender }
