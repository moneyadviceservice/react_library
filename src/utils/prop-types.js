import PropTypes from 'prop-types'
import { SPACING_SIZES } from './constants'

/** spacing prop.type */
const spacingProp = PropTypes.shape({
  bottom: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
  horizontal: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
  left: PropTypes.oneOfType([PropTypes.oneOf(SPACING_SIZES), PropTypes.string]),
  right: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
  top: PropTypes.oneOfType([PropTypes.oneOf(SPACING_SIZES), PropTypes.string]),
  vertical: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
})

export const genericPropTypes = {
  a11yTitle: PropTypes.string,
  alignSelf: PropTypes.oneOf(['start', 'center', 'end', 'stretch']),
  margin: PropTypes.oneOfType([
    PropTypes.oneOf(['none', ...SPACING_SIZES]),
    spacingProp,
    PropTypes.string,
    PropTypes.object,
  ]),
  padding: PropTypes.oneOfType([
    PropTypes.oneOf([...SPACING_SIZES]),
    spacingProp,
    PropTypes.string,
    PropTypes.object,
  ]),
  border: PropTypes.oneOfType([
    PropTypes.oneOf(['none', ...SPACING_SIZES]),
    spacingProp,
    PropTypes.string,
    PropTypes.object,
  ]),
}

export function genericPropsDefaults({
  a11yTitle = null,
  alignSelf = null,
  margin = null,
  padding = null,
  border = null,
} = {}) {
  return {
    a11yTitle: a11yTitle,
    alignSelf: alignSelf,
    margin: margin,
    padding: padding,
    border: border,
  }
}
