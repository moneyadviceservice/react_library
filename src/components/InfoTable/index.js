import React from 'react'
import PropTypes from 'prop-types'
import { genericPropTypes, genericPropsDefaults } from '../../utils/prop-types'
import {
  TableContainer,
  TableIcon,
  TableHead,
  TableContent,
  TableText,
} from './StyledInfoTable'
import QuestionIcon from '../../assets/Images/question_mark.svg'

const InfoTable = ({
  a11yTitle,
  children,
  icon,
  padding,
  tableColor,
  textContent,
  title,
  titleColor,
  ...rest
}) => (
  <TableContainer
    a11yTitle={a11yTitle || title}
    tableColor={tableColor}
    {...rest}
  >
    <TableHead
      direction="row"
      align="center"
      flexWrap="nowrap"
      tableColor={tableColor}
      titleColor={titleColor}
    >
      <TableIcon titleColor={titleColor}>
        {icon ? icon : <QuestionIcon />}
      </TableIcon>
      {title}
    </TableHead>
    <TableContent padding={padding}>
      {textContent && <TableText>{textContent}</TableText>}
      {children}
    </TableContent>
  </TableContainer>
)

// Documentation
InfoTable.propTypes = {
  /** Text inside the Label component. Works as the caption for the checkbox element. */
  children: PropTypes.any,
  /** Provide a custom svg icon to the table header. */
  icon: PropTypes.node,
  /** Sets a custom color to the table and border. */
  tableColor: PropTypes.string,
  /** Text content to be placed inside the table. */
  textContent: PropTypes.string,
  /** Title to be placed in table header. */
  title: PropTypes.string,
  /** Sets a custom color to the svg icon and title text in table header. */
  titleColor: PropTypes.string,
  ...genericPropTypes,
}

InfoTable.defaultProps = {
  titleColor: 'white',
  ...genericPropsDefaults(),
}

/** @component */
export { InfoTable }
