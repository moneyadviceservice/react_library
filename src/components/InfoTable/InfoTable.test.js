import React from 'react'
import { render } from '../../utils/testing'
import { InfoTable } from '../InfoTable'

describe('<InfoTable />', () => {
  describe('Snapshots', () => {
    it('should match snapshot with title and textContent props', () => {
      const { container, getByTestId } = render(
        <InfoTable title="example" textContent="example content" />
      )
      // snapshot
      expect(container).toMatchSnapshot()
      // content
      expect(getByTestId('content').textContent).toBe('example content')
    })
    it('should match snapshot with title prop and children', () => {
      const { container, getByTestId } = render(
        <InfoTable title="example">example child</InfoTable>
      )
      // snapshot
      expect(container).toMatchSnapshot()
      // content
      expect(getByTestId('content').textContent).toBe('example child')
    })
    it('should match snapshot with tableColor and titleColor props', () => {
      const { container } = render(
        <InfoTable
          title="example"
          textContent="example content"
          tableColor="aqua"
          titleColor="black"
        />
      )
      expect(container).toMatchSnapshot()
    })
  })
})
