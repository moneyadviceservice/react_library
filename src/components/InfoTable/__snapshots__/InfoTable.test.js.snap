// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InfoTable /> Snapshots should match snapshot with tableColor and titleColor props 1`] = `
<div>
  <div
    aria-label="example"
    class="styledCol__ColWrapper-sc-1mivnjm-0 fhguMU StyledInfoTable__TableContainer-t9zemp-0 iRNxLK"
  >
    <h3
      class="styledCol__ColWrapper-sc-1mivnjm-0 cJqMBd StyledInfoTable__TableHead-t9zemp-2 dWFYiL"
    >
      example
    </h3>
    <div
      class="styledCol__ColWrapper-sc-1mivnjm-0 fhguMU StyledInfoTable__TableContent-t9zemp-3 esOBTz"
      data-testid="content"
    >
      <p
        class="StyledParagraph-sc-1toaej-0 gpltyO StyledInfoTable__TableText-t9zemp-4"
      >
        example content
      </p>
    </div>
  </div>
</div>
`;

exports[`<InfoTable /> Snapshots should match snapshot with title and textContent props 1`] = `
<div>
  <div
    aria-label="example"
    class="styledCol__ColWrapper-sc-1mivnjm-0 fhguMU StyledInfoTable__TableContainer-t9zemp-0 egTFuN"
  >
    <h3
      class="styledCol__ColWrapper-sc-1mivnjm-0 cJqMBd StyledInfoTable__TableHead-t9zemp-2 ekcFyQ"
    >
      example
    </h3>
    <div
      class="styledCol__ColWrapper-sc-1mivnjm-0 fhguMU StyledInfoTable__TableContent-t9zemp-3 esOBTz"
      data-testid="content"
    >
      <p
        class="StyledParagraph-sc-1toaej-0 gpltyO StyledInfoTable__TableText-t9zemp-4"
      >
        example content
      </p>
    </div>
  </div>
</div>
`;

exports[`<InfoTable /> Snapshots should match snapshot with title prop and children 1`] = `
<div>
  <div
    aria-label="example"
    class="styledCol__ColWrapper-sc-1mivnjm-0 fhguMU StyledInfoTable__TableContainer-t9zemp-0 egTFuN"
  >
    <h3
      class="styledCol__ColWrapper-sc-1mivnjm-0 cJqMBd StyledInfoTable__TableHead-t9zemp-2 ekcFyQ"
    >
      example
    </h3>
    <div
      class="styledCol__ColWrapper-sc-1mivnjm-0 fhguMU StyledInfoTable__TableContent-t9zemp-3 esOBTz"
      data-testid="content"
    >
      example child
    </div>
  </div>
</div>
`;
